scripts/igmp_query_injector.py

```diff
@@ -1,267 +1,62 @@
-        return struct.pack('!%ds' % len(s), s)
-
-    def create_igmp_layer(self):
-        """
-        IGMP Version: 2
-        IGMP Type: IGMP Query (0x11)
-        Max Resp Time: self.max_resp_time
-        Multicast Address: 0.0.0.0
-        """
-        msg_type = 0x11
-        # MAX_RESP_TIME filed unit is 100ms
-        max_resp_time = self.max_resp_time / 100
-        group_address = socket.inet_aton('0.0.0.0')
-        chksum = self._calc_igmp_checksum(struct.pack('!BBH4s', msg_type, max_resp_time, 0x0000, group_address))
-        return struct.pack('!BBH4s', msg_type, max_resp_time, chksum, group_address)
-
-    def _calc_igmp_checksum(self, packet):
-        def carry_around_add(a, b):
-            c = a + b
-            return (c & 0xffff) + (c >> 16)
-
-        s = 0
-        msg = struct.unpack('%dB' % len(packet), packet)
-        for i in range(0, len(msg), 2):
-            w = (msg[i] << 8) + msg[i+1]
-            s = carry_around_add(s, w)
-        return ~s & 0xffff
-
-    def generate(self):
-        ether_layer = self.create_ether_layer()
-        vlan_layer = self.create_vlan_layer()
-        ip_layer = self.create_ip_layer()
-        igmp_layer = self.create_igmp_layer()
-        packet = ether_layer + vlan_layer + ip_layer + igmp_layer
-        return packet
+    def read_watch(self):
+        return g_xs_handler.read_watch()
 
 
 class IGMPQueryInjector(object):
-    __metaclass__ = ABCMeta
-
-    def __init__(self, max_resp_time):
-        self.max_resp_time = max_resp_time
-
-    def inject_to_vif(self, vif, mac):
-        log.info('Inject IGMP query to vif:%s, mac:%s' % (vif, mac))
-        packet = IGMPQueryGenerator(mac, -1, self.max_resp_time).generate()
-        self.inject_query_packet(vif, packet)
-
-    def inject_to_vifs(self, vifs):
-        for vif in vifs:
-            _vif = Vif(vif)
-            self.inject_to_vif(_vif.vif_name, _vif.mac)
-
-    def inject_query_packet(self, interface, packet):
-        s = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, 0)
-        s.bind((interface, 0))
-        s.send(packet)
-        s.close()
-
-    @abstractmethod
-    def inject(self):
-        pass
-
-
-class VifsInjector(IGMPQueryInjector):
     def __init__(self, max_resp_time, vifs, vif_connected_timeout=0):
-        super(VifsInjector, self).__init__(max_resp_time)
+        self.max_resp_time = max_resp_time
         self.vifs = vifs
         self.vif_connected_timeout = vif_connected_timeout
 
+    def inject_to_vif(self, vif):
+        mac = get_vif_mac(vif)
+        try:
+            self.inject_packet(vif, mac)
+            log.info('Inject IGMP query to vif:%s, mac:%s' % (vif, mac))
+        except:
+            log.error('Inject IGMP query to vif:%s, mac:%s failed' % (vif, mac))
+
+    def inject_packet(self, iface, dst_mac):
+        ether_part = Ether(src='00:00:00:00:00:00', dst=dst_mac)
+        ip_part = IP(ttl=1, src='0.0.0.0', dst='224.0.0.1')
+        igmp_part = IGMP(type=0x11)
```

Comment:
We will not be able to make a new EPEL snapshot until Monday when Artifactory is upgraded, so you might have to put the definition of the IGMP class back into this script temporarily.  Sorry!